### PR TITLE
Moved reporter to separate package for reuse.

### PR DESCRIPTION
**Pull request by Bplotka**
Additionally:
  - http_prometheus has now it's own Middleware/Tripperware constructor

PTAL @devnev 
Signed-off-by: Bartek Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: reporter/middleware.go [hunk 1]
```diff
@@ -10,10 +10,10 @@ import (
 	"github.com/mwitkow/go-httpwares"
 )
 
-// Middleware returns a http.Handler middleware that exports request metrics.
+// Middleware returns a http.Handler middleware that set up reporter callbacks.
 // If the tags middleware is used, this should be placed after tags to pick up metadata.
 // This middleware assumes HTTP/1.x-style requests/response behaviour. It will not work with servers that use
-// hijacking, pushing, or other similar features.
+// hijacking, pushing, or other similar features. (TODO)
```
Comment (devnev) [inline reporter/middleware.go]: Drop the TODO, open an Issue if it's something you're going to work on.

File: metrics/doc.go [hunk 2]
```diff
@@ -1,12 +0,0 @@
...
-
-/*
-`http_metrics` provides client and server side reporting of HTTP stats.
-
-The middleware (server-side) or tripperware (client-side) must be given a reporter to record the stats for each request.
-
-Prometheus-based reporter implementations for client and server metrics are included. The user may choose what level of
-detail is included using options to these reporters.
-*/
-package http_metrics
```
Comment (devnev) [inline metrics/doc.go]: if you're removing all the stuff in the metrics package, i think the two levels stop making sense. move the prometheus subpackage up.

File: metrics/prometheus/client.go [hunk 3]
```diff
@@ -78,11 +79,18 @@ func ClientMetrics(opts ...opt) http_metrics.Reporter {
 	return &clientReporter{opts: o}
 }
 
+// Tripperware returns a new client-side ware that exports request metrics.
+// It is using Reporter Tripperware with Prometheus Client Metrics Reporter.
```
Comment (devnev) [inline metrics/prometheus/client.go]: Reporter is an implementation detail

File: reporter/doc.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
+// Copyright 2017 Mark Nevill. All Rights Reserved.
+// See LICENSE for licensing terms.
+
+/*
+`reporter` provides client and server side reporting of HTTP request and response for stats.
```
Comment (devnev) [inline reporter/doc.go]: this entire doc still reads like it's for metrics. the reporter doesn't do anything with stats, its provides callback for particular request-processing events

File: reporter/middleware.go [hunk 5]
```diff
@@ -10,10 +10,10 @@ import (
 	"github.com/mwitkow/go-httpwares"
 )
 
-// Middleware returns a http.Handler middleware that exports request metrics.
+// Middleware returns a http.Handler middleware that set up reporter callbacks.
```
Comment (devnev) [inline reporter/middleware.go]: "that sets up reporter callbacks"

File: reporter/tripperware.go [hunk 6]
```diff
@@ -11,7 +11,7 @@ import (
 	"github.com/mwitkow/go-httpwares"
 )
 
-// Tripperware returns a new client-side ware that exports request metrics.
+// Tripperware returns a new client-side ware that have reporter callbacks set up.
```
Comment (devnev) [inline reporter/tripperware.go]: "that has reporter callbacks set up"

File: reporter/tripperware.go [hunk 7]
```diff
@@ -11,7 +11,7 @@ import (
...
 // If the tags tripperware is used, this should be placed after tags to pick up metadata.
```
Comment (devnev) [inline reporter/tripperware.go]: this tags comment isn't relevant to the reporters

File: metrics/doc.go (hunk 2 above)
Comment (Bplotka) [reply metrics/doc.go]: Hm, again it is consistent with logrus now. But I agree, we will not have plans for any othere metrics, so will do

File: metrics/prometheus/client.go (hunk 3 above)
Comment (Bplotka) [reply metrics/prometheus/client.go]: Killed

File: reporter/doc.go (hunk 4 above)
Comment (Bplotka) [reply reporter/doc.go]: Well it is `for stats` because it enables stats (: Will rephrase.

File: metrics/doc.go (hunk 2 above)
Comment (Bplotka) [reply metrics/doc.go]: Done

**Pull request closed without merging**